File: android/app/src/main/java/com/twintipsolutions/aclrehabtracker/data/service/GeminiPoseService.kt
package com.twintipsolutions.aclrehabtracker.data.service

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.util.Base64
import android.util.Log
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.functions.FirebaseFunctions
import com.twintipsolutions.aclrehabtracker.data.model.Keypoint
import com.twintipsolutions.aclrehabtracker.data.model.PoseResult
import com.google.firebase.functions.FirebaseFunctionsException
import kotlinx.coroutines.tasks.await
import java.io.ByteArrayOutputStream
import java.io.File
import kotlin.math.max
import kotlin.math.roundToInt

object GeminiPoseService {
    private val functions = FirebaseFunctions.getInstance("us-central1")

    suspend fun detectKneeAngle(
        imagePath: String,
        injuredKnee: String? = null,
        injuryType: String? = null
    ): PoseResult {
        val bitmap = BitmapFactory.decodeFile(imagePath)
            ?: throw Exception("Could not load image from path")
        return detectKneeAngle(bitmap, injuredKnee, injuryType)
    }

    suspend fun detectKneeAngle(
        bitmap: Bitmap,
        injuredKnee: String? = null,
        injuryType: String? = null
    ): PoseResult {
        val resized = resizeImage(bitmap, 1024)
        val base64 = bitmapToBase64(resized)

        val data = hashMapOf<String, Any>(
            "imageBase64" to base64
        )
        injuredKnee?.let { data["injuredKnee"] = it }
        injuryType?.let { data["injuryType"] = it }

        // Ensure user is authenticated before calling Cloud Function
        var user = FirebaseAuth.getInstance().currentUser
        if (user == null) {
            Log.d("GeminiPoseService", "No auth user, attempting sign-in...")
            try {
                AuthService.signInAnonymously()
                user = FirebaseAuth.getInstance().currentUser
            } catch (e: Exception) {
                Log.e("GeminiPoseService", "Auth sign-in failed: ${e.message}", e)
            }
        }
        Log.d("GeminiPoseService", "Auth state: uid=${user?.uid}, isAnon=${user?.isAnonymous}, base64Size=${base64.length}")

        val result = try {
            functions
                .getHttpsCallable("analyzeKneeAngle")
                .call(data)
                .await()
        } catch (e: FirebaseFunctionsException) {
            Log.e("GeminiPoseService", "FirebaseFunctionsException: code=${e.code}, message=${e.message}, details=${e.details}", e)
            throw when (e.code) {
                FirebaseFunctionsException.Code.UNAUTHENTICATED ->
                    Exception("Authentication error. Please restart the app and try again.")
                FirebaseFunctionsException.Code.UNAVAILABLE ->
                    Exception("AI service is temporarily unavailable. Please try again later.")
                FirebaseFunctionsException.Code.DEADLINE_EXCEEDED ->
                    Exception("Analysis took too long. Please try again with a clearer photo.")
                else ->
                    Exception("Could not analyze the photo. Please try again.")
            }
        } catch (e: Exception) {
            Log.e("GeminiPoseService", "Exception: ${e.javaClass.name}: ${e.message}", e)
            throw Exception("Could not analyze the photo. Please check your connection and try again.")
        }

        @Suppress("UNCHECKED_CAST")
        val responseData = result.getData() as Map<String, Any>

        val rawAngle = (responseData["angle"] as? Number)?.toInt()
            ?: throw Exception("AI could not detect knee angle. Please try repositioning your leg and take another photo.")
        val angle = rawAngle.coerceIn(0, 180)
        val confidence = (responseData["confidence"] as? Number)?.toDouble() ?: 0.0

        @Suppress("UNCHECKED_CAST")
        val hipData = responseData["hip"] as? Map<String, Any>
        @Suppress("UNCHECKED_CAST")
        val kneeData = responseData["knee"] as? Map<String, Any>
        @Suppress("UNCHECKED_CAST")
        val ankleData = responseData["ankle"] as? Map<String, Any>

        return PoseResult(
            hip = parseKeypoint(hipData),
            knee = parseKeypoint(kneeData),
            ankle = parseKeypoint(ankleData),
            angle = angle,
            confidence = confidence
        )
    }

    private fun parseKeypoint(data: Map<String, Any>?): Keypoint {
        if (data == null) return Keypoint(0f, 0f)
        return Keypoint(
            x = (data["x"] as? Number)?.toFloat() ?: 0f,
            y = (data["y"] as? Number)?.toFloat() ?: 0f
        )
    }

    private fun resizeImage(bitmap: Bitmap, maxDimension: Int): Bitmap {
        val width = bitmap.width
        val height = bitmap.height
        val maxDim = max(width, height)
        if (maxDim <= maxDimension) return bitmap

        val scale = maxDimension.toFloat() / maxDim
        val newWidth = (width * scale).roundToInt()
        val newHeight = (height * scale).roundToInt()
        return Bitmap.createScaledBitmap(bitmap, newWidth, newHeight, true)
    }

    private fun bitmapToBase64(bitmap: Bitmap): String {
        val baos = ByteArrayOutputStream()
        bitmap.compress(Bitmap.CompressFormat.JPEG, 80, baos)
        return Base64.encodeToString(baos.toByteArray(), Base64.NO_WRAP)
    }
}

File: ios/ios_new/ACLRehabTracker/Services/GeminiPoseService.swift
import Foundation
import UIKit
import FirebaseFunctions

class GeminiPoseService {
    static let shared = GeminiPoseService()

    private lazy var functions = Functions.functions(region: "us-central1")

    private init() {}

    /// Detect knee angle from an image using Vertex AI via Cloud Functions
    /// - Parameters:
    ///   - image: UIImage of the leg
    ///   - injuredKnee: Which knee is injured (left/right) - helps AI identify correct leg
    ///   - injuryType: Type of injury for context
    /// - Returns: PoseResult with the detected angle
    func detectKneeAngle(image: UIImage, injuredKnee: KneeSide? = nil, injuryType: InjuryType? = nil) async throws -> PoseResult {
        // Resize image if too large
        let resizedImage = resizeImageIfNeeded(image, maxDimension: 1024)

        // Convert to base64
        guard let imageData = resizedImage.jpegData(compressionQuality: 0.8) else {
            throw GeminiPoseError.imageLoadFailed
        }
        let base64String = imageData.base64EncodedString()

        // Build request data
        var requestData: [String: Any] = ["imageBase64": base64String]

        // Add injury info if available
        if let knee = injuredKnee {
            requestData["injuredKnee"] = knee.rawValue
        }
        if let injury = injuryType {
            requestData["injuryType"] = injury.rawValue
        }

        // Call the Cloud Function
        do {
            let callable = functions.httpsCallable("analyzeKneeAngle")
            let result = try await callable.call(requestData)

            guard let data = result.data as? [String: Any] else {
                throw GeminiPoseError.invalidResponse
            }

            return try parseCloudFunctionResponse(data)
        } catch let error as GeminiPoseError {
            throw error
        } catch let error as NSError {
            // Handle Firebase Functions errors with user-friendly messages
            if error.domain == FunctionsErrorDomain {
                let code = FunctionsErrorCode(rawValue: error.code)
                switch code {
                case .unauthenticated:
                    throw GeminiPoseError.detectionFailed("Authentication error. Please restart the app and try again.")
                case .unavailable:
                    throw GeminiPoseError.detectionFailed("AI service is temporarily unavailable. Please try again later.")
                case .deadlineExceeded:
                    throw GeminiPoseError.detectionFailed("Analysis took too long. Please try again with a clearer photo.")
                default:
                    throw GeminiPoseError.detectionFailed("Could not analyze the photo. Please try again.")
                }
            }
            throw GeminiPoseError.detectionFailed("Could not analyze the photo. Please check your connection and try again.")
        }
    }

    /// Detect knee angle from an image file path
    func detectKneeAngle(imagePath: String, injuredKnee: KneeSide? = nil, injuryType: InjuryType? = nil) async throws -> PoseResult {
        guard let image = UIImage(contentsOfFile: imagePath) else {
            throw GeminiPoseError.imageLoadFailed
        }
        return try await detectKneeAngle(image: image, injuredKnee: injuredKnee, injuryType: injuryType)
    }

    // MARK: - Private Helpers

    private func resizeImageIfNeeded(_ image: UIImage, maxDimension: CGFloat) -> UIImage {
        let size = image.size

        guard size.width > maxDimension || size.height > maxDimension else {
            return image
        }

        let scale = maxDimension / max(size.width, size.height)
        let newSize = CGSize(width: size.width * scale, height: size.height * scale)

        UIGraphicsBeginImageContextWithOptions(newSize, false, 1.0)
        image.draw(in: CGRect(origin: .zero, size: newSize))
        let resizedImage = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()

        return resizedImage ?? image
    }

    private func parseCloudFunctionResponse(_ data: [String: Any]) throws -> PoseResult {
        guard let angle = data["angle"] as? Int,
              let confidence = data["confidence"] as? Double,
              let hip = data["hip"] as? [String: Double],
              let knee = data["knee"] as? [String: Double],
              let ankle = data["ankle"] as? [String: Double],
              let hipX = hip["x"], let hipY = hip["y"],
              let kneeX = knee["x"], let kneeY = knee["y"],
              let ankleX = ankle["x"], let ankleY = ankle["y"] else {
            throw GeminiPoseError.invalidResponse
        }

        let clampedAngle = max(0, min(180, angle))

        return PoseResult(
            hip: Keypoint(x: hipX, y: hipY, confidence: Float(confidence)),
            knee: Keypoint(x: kneeX, y: kneeY, confidence: Float(confidence)),
            ankle: Keypoint(x: ankleX, y: ankleY, confidence: Float(confidence)),
            angle: clampedAngle
        )
    }
}

// MARK: - Errors

enum GeminiPoseError: LocalizedError {
    case imageLoadFailed
    case noResponse
    case invalidResponse
    case detectionFailed(String)

    var errorDescription: String? {
        switch self {
        case .imageLoadFailed:
            return "Failed to load image"
        case .noResponse:
            return "No response from server"
        case .invalidResponse:
            return "Could not parse server response"
        case .detectionFailed(let message):
            return message
        }
    }
}
